### PR TITLE
Fix block vacancy bug

### LINKED ISSUE #4695: KeepUp Node panics at switch block with load
RC1 Testing:

KeepUp Nodes shutting down at/near switch block.

Network: 5 + 2
Name: ajith-condor-small (8s)
Branch: feat-2.0
Commit: fe996ee6bf859afec1925f01baf60f900bae6a77
Config: {"config":{}, "chainspec":{"era_duration": \"30min\", "minimum_block_time": \"8192 ms\", "block_max_mint_count": 325, "block_max_auction_count": 50, "block_max_standard_count": 15}}
Genesis: feat-2.0 (fe996ee6bf859afec1925f01baf60f900bae6a77)
Test: Transfer and WASM Load Testing
30k Transfers
2.6k WASM Transfers
:red_circle: Issue: KeepUp Node panics at switch block with load
Observations:

- network running in 8s blocks

- KeepUp Nodes shutting down at/near switch block

- when active loading/executing transactions near to switch block

- node recovers to KeepUp after era change

- sidecar rpc service was impacted

- errors in log

"message":"dropped without being responded to outside of shutdown","responder":"Responder<()>"
"message":"request for channel closed, this may be a bug? check if a component is stuck from now on","err":"oneshot canceled","channel":"\"()\""},"target":"casper_node::effect"
"message":"fatal error via control announcement","file":"node/src/components/storage.rs","line":"230","msg":"storage error: duplicate entries for blocks for transaction deploy-hash(8da2..9081): block-hash(257f..c272) / block-hash(d939..9270)"
Validating nodes in healthy state
avg finalisation times of 12ms
avg finalisation times on load 225ms
no missed proposals were observed
always in 'validate' state
Grafana: [link](https://grafana.casperlabs.io/d/rpvdF91Sk/release-test-metrics-overview-condor?orgId=1&var-casper_environment=ajith-condor-small&var-casper_version=casper-ajith-condor-small&var-casper_node_type=All&var-casper_hostname=All&from=1714525576870&to=1714527002090)
http://genesis.casperlabs.io/ajith-condor-small/casper-node-dumps/ajith-condor-small/01052024_0123/172.142.93.244_sidecar.log.tar.gz

### LINKED ISSUE #4696: Emergency Upgrade - Network stalls.
RC1 Testing:

> Upgrade passed But the network stalled after about 20 blocks

Test Execution:

Network: 5 + 2
Name: jiuhong-test-jh
Genesis: feat-1.5.6 (9f3995853204a18f17de9c022233d22aa14b9c37)
Upgrade: 1.5.6 -> 2.0.0  (d6cf1e7a6b93db178751711394809ac5351ce3ff)
Test:
Emergency Restart (with validator rotation)

* Shutdown all nodes 
* Wait for 1 era
* Start all nodes back up with emergency restart.
* The network should continue.
* Startup nodes removed from new validator list
More testing scenario:
All the 7 nodes were shutdown.
Wait for 135 minutes.
Start the new version 2_0_0.
Then after around 20 blocks the network stalled.
:red_circle: Observations:
Upgrade passed But the network stalled after about 20 blocks
http://genesis.casperlabs.io/jiuhong-test-jh/casper-node-dumps/jiuhong-test-jh/02052024_0830/dump_download_list.html
The log shows
{"timestamp":"2024-05-02T14:46:09.817684Z","level":"ERROR","fields":{"message":"distribute block rewards failed due to auction error ValidatorNotFound"},"target":"casper_storage::global_state::state"}
{"timestamp":"2024-05-02T14:46:09.817807Z","level":"ERROR","fields":{"message":"failed to execute block","error":"Auction error: Validator not found"},"target":"casper_node::components::contract_runtime::utils"}

### LINKED ISSUE #4698: Failed to start network with Mainnet DB snapshot
RC Testing:

Network: 5 + 2
Name: ajith-mainnet-clone
Branch: feat-2.0
Commit: 47cd7d31c38f51bd1c646d7a2948fec5f12749d9
Switch Block: 2163794
Test: Mainnet DB Snapshot
:red_circle: Issue: Failed to start network with Mainnet DB snapshot
Observations:
used the feat-2.0 global-state-update-gen tool
used db snapshot from era 11242 and height 2163794
getting error with commit upgrade
{"timestamp":"2024-05-06T11:34:03.487763Z","level":"ERROR","fields":{"message":"fatal error via control announcement","file":"node/src/reactor/main_reactor/control.rs","line":"125","msg":"failed to commit upgrade: Value not found"},"target":"casper_node::reactor","span":{"ev":1572,"name":"dispatch"},"spans":[{"ev":1572,"name":"dispatch"}]}
Grafana: [link](https://grafana.casperlabs.io/d/rpvdF91Sk/release-test-metrics-overview-condor?orgId=1&var-casper_environment=ajith-mainnet-clone&var-casper_version=casper&var-casper_node_type=All&var-casper_hostname=All&from=1714994563898&to=1714995401346)
Dumps: http://genesis.casperlabs.io/ajith-mainnet-clone/casper-node-dumps/ajith-mainnet-clone/06052024_1130/dump_download_list.html

### PR DESCRIPTION
**Pull request by darthsiroftardis**
CHANGELOG:

- Fix bug in calculating `next_era_gas_price` for non-validating nodes
- Fix bug in migrating of contract and packages
- Fix bug in global state update during protocol upgrade

Closes #4695
Closes #4696
Closes #4698

### CODE REVIEW

File: execution_engine/src/runtime_context/mod.rs [hunk 1]
```diff
@@ -882,12 +883,12 @@ where
 
     pub(crate) fn migrate_contract(
```
Comment (fizyk20) [inline execution_engine/src/runtime_context/mod.rs]: How about renaming this to `migrate_package`, too?

File: storage/src/tracking_copy/ext_entity.rs [hunk 2]
```diff
@@ -566,113 +566,109 @@ where
...
+            let entity_hash = AddressableEntityHash::new(contract_hash.value());
+            let entity_key = Key::contract_entity_key(entity_hash);
+            self.write(entity_key, StoredValue::AddressableEntity(updated_entity));
+
+            package.insert_entity_version(protocol_version.value().major, entity_hash);
         }
 
+        let access_key_value = CLValue::from_t(access_uref).map_err(Self::Error::CLValue)?;
+
+        let pacakage_key = Key::Package(
```
Comment (fizyk20) [inline storage/src/tracking_copy/ext_entity.rs]:
```suggestion
        let package_key = Key::Package(
```

**Review by fizyk20 — APPROVED** [review]
LGTM (only some naming nitpicks) :+1:

**Review by EdHastingsCasperAssociation — APPROVED** [review]

**Pull request merged**